materialize-bigquery/bigquery.go

```diff
@@ -1,214 +1,87 @@
 package main
 
 import (
-	"context"
-	"encoding/json"
+	"errors"
 	"fmt"
-	"regexp"
 
 	"cloud.google.com/go/bigquery"
-	storage "cloud.google.com/go/storage"
-	boilerplate "github.com/estuary/connectors/materialize-boilerplate"
 	pf "github.com/estuary/flow/go/protocols/flow"
-	pm "github.com/estuary/flow/go/protocols/materialize"
-	sqlDriver "github.com/estuary/flow/go/protocols/materialize/sql"
-	log "github.com/sirupsen/logrus"
-	"google.golang.org/api/option"
 )
 
-func main() {
-	boilerplate.RunMain(newBigQueryDriver())
-}
-
-// Config represents the endpoint configuration for BigQuery.
-type config struct {
-	BillingProjectID string `json:"billing_project_id,omitempty"`
-	ProjectID        string `json:"project_id"`
-	Dataset          string `json:"dataset"`
-	Region           string `json:"region,omitempty"`
-	Bucket           string `json:"bucket"`
-	BucketPath       string `json:"bucket_path"`
-	CredentialsFile  string `json:"credentials_file,omitempty"`
-	CredentialsJSON  []byte `json:"credentials_json,omitempty"`
-}
-
-func (c *config) Validate() error {
-	if c.ProjectID == "" {
-		return fmt.Errorf("expected project_id")
-	}
-	if c.Dataset == "" {
-		return fmt.Errorf("expected dataset")
-	}
-	if c.Region == "" {
-		return fmt.Errorf("expected region")
-	}
-	if c.Bucket == "" {
-		return fmt.Errorf("expected bucket")
-	}
-	return nil
-}
-
-// DatasetPath returns the sqlDriver.ResourcePath including the dataset.
-func (c *config) DatasetPath(path ...string) sqlDriver.ResourcePath {
-	return append([]string{c.ProjectID, c.Dataset}, path...)
-}
-
-type tableConfig struct {
-	base *config
-
-	Table string `json:"table"`
-	Delta bool   `json:"delta_updates,omitempty"`
-}
+var (
+	errPrivateProjection = errors.New("projection is private and should not be included")
+	errNotFound          = errors.New("not found")
+)
 
-func (c *tableConfig) Validate() error {
-	if c.Table == "" {
-		return fmt.Errorf("expected table")
+// Because binding.Collection.Projections doesn't respect the same order as
+// binding.FieldSelection.AllFields(), the schema needs to be generated by having an inner loop
+// matching the projection then assigning it the index as defined in AllFields().
+// If binding.Collection.Projections would honor the same ordering, this function could
+// be simplified to just iterate over the projection slice and remove the need for a few
+// extra error handling as defined in fieldSchemaForNameAndProjections
+func schemaForBinding(binding *pf.MaterializationSpec_Binding) (bigquery.Schema, error) {
+	fields := binding.FieldSelection.AllFields()
+	schema := make(bigquery.Schema, len(fields))
+
+	for idx, fieldName := range fields {
+		var (
+			err   error
+			field *bigquery.FieldSchema
+		)
+
+		if field, err = fieldSchemaForNameAndProjections(fieldName, binding.Collection.Projections); err == errPrivateProjection {
```

Comment:
Maybe I'm missing something but I couldn't find any place where this error is emitted

Reply:
I mean `errPrivateProjection`. `fieldSchemaForNameAndProjections` doesn't seem to emit such an error ever

Reply:
It is on line 36 because it only `continue` if the err is equal to `errPrivatePRojection` but since the both of you tripped on that line, I'll change it to make it clearer (also you said in another comment that this might not even be needed, so maybe I'll just remove it.